compiler/fir/tree/src/org/jetbrains/kotlin/fir/types/FirTypeUtils.kt

```diff
@@ -138,3 +138,19 @@ private fun ConeTypeParameterType.hasNotNullUpperBound(): Boolean {
         }
     }
 }
+
+fun FirTypeRef.containsNull(): Boolean =
+    coneType.containsNull()
+
+fun ConeKotlinType.containsNull(): Boolean {
+    if (isMarkedNullable) {
+        return true
+    }
+    return when (this) {
+        is ConeFlexibleType -> upperBound.containsNull()
+        is ConeDefinitelyNotNullType -> false
+        is ConeTypeParameterType -> true
```

Comment:
Shouldn't it be dependent to parameter bound type? For example `T : Any` cannot be null.

Reply:
Oh, you're right, I was too naive here. Will update it as per suggestion.

Reply:
Rewrote this to consider parameter bound types.